File: apps/web/src/lib/api/performance.ts
/**
 * Performance API -- content analytics, voice analysis, and drift detection.
 */

import { apiFetch } from "./client";

// ── Content Post Types ───────────────────────────────────

export interface ContentPostSummary {
  id: string;
  title: string;
  content_type: string;
  platform: string;
  hook_type: string | null;
  topic: string | null;
  topic_category: string | null;
  performance_tier: string | null;
  engagement_rate: number | null;
  views: number | null;
  likes: number | null;
  comments: number | null;
  published_at: string | null;
  created_at: string;
}

export interface ContentPostDetail extends ContentPostSummary {
  hook_used: string | null;
  content_body: string | null;
  workflow_id: string | null;
  collection_id: string | null;
  published_url: string | null;
  day_of_week: string | null;
  shares: number | null;
  saves: number | null;
  watch_time_seconds: number | null;
  click_through_rate: number | null;
  impressions: number | null;
  reach: number | null;
  subscribers_gained: number | null;
  agent_analysis: Record<string, any>;
  tags: string[];
  metadata: Record<string, any>;
  updated_at: string;
}

// ── Analytics Types ──────────────────────────────────────

export interface PlatformBreakdown {
  platform: string;
  post_count: number;
  avg_engagement_rate: number | null;
  avg_views: number | null;
  top_tier_count: number;
}

export interface TopicBreakdown {
  topic_category: string;
  post_count: number;
  avg_engagement_rate: number | null;
  avg_views: number | null;
}

export interface HookBreakdown {
  hook_type: string;
  post_count: number;
  avg_engagement_rate: number | null;
  example_hooks: string[];
}

export interface PatternDetected {
  pattern: string;
  evidence: string;
  confidence: number;
}

export interface PerformanceAnalytics {
  total_posts: number;
  platforms: PlatformBreakdown[];
  top_topics: TopicBreakdown[];
  top_hook_types: HookBreakdown[];
  best_day_of_week: string | null;
  patterns: PatternDetected[];
  top_hooks: string[];
  anti_hooks: string[];
}

// ── Performance API ──────────────────────────────────────

export const performanceApi = {
  list: (platform?: string, tier?: string, brandId?: string) => {
    const params = new URLSearchParams();
    if (platform) params.set("platform", platform);
    if (tier) params.set("tier", tier);
    if (brandId) params.set("brand_id", brandId);
    const qs = params.toString();
    return apiFetch<ContentPostSummary[]>(
      `/content-posts${qs ? `?${qs}` : ""}`
    );
  },

  get: (id: string) => apiFetch<ContentPostDetail>(`/content-posts/${id}`),

  create: (data: {
    title: string;
    content_type: string;
    platform: string;
    hook_used?: string;
    hook_type?: string;
    topic?: string;
    topic_category?: string;
    content_body?: string;
    published_url?: string;
    published_at?: string;
    day_of_week?: string;
    tags?: string[];
    brand_id?: string;
  }) =>
    apiFetch<ContentPostSummary>("/content-posts", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  updateMetrics: (
    id: string,
    data: {
      views?: number;
      likes?: number;
      comments?: number;
      shares?: number;
      saves?: number;
      watch_time_seconds?: number;
      click_through_rate?: number;
      impressions?: number;
      reach?: number;
      subscribers_gained?: number;
    }
  ) =>
    apiFetch<ContentPostDetail>(`/content-posts/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    }),

  analytics: (brandId?: string) => {
    const qs = brandId ? `?brand_id=${brandId}` : "";
    return apiFetch<PerformanceAnalytics>(`/content-posts/analytics${qs}`);
  },

  analyze: (id: string) =>
    apiFetch<{
      post_id: string;
      performance_tier: string | null;
      analysis: Record<string, any>;
      message: string;
    }>(`/content-posts/${id}/analyze`, { method: "POST" }),
};

// ── Voice / Drift Types ──────────────────────────────────

export interface SelfVoiceDNA {
  tone: string;
  sentence_style: string;
  vocabulary_level: string;
  avg_sentence_length: number | null;
  hook_patterns: string[];
  cta_patterns: string[];
  signature_phrases: string[];
  content_structure: string;
  personality_traits: string[];
  sample_hooks: string[];
  posts_analyzed: number;
}

export interface VoiceDriftResult {
  drift_score: number;
  drift_level: string;
  details: string[];
  recommendation: string;
  baseline_available: boolean;
}

// ── Voice API ────────────────────────────────────────────

export const voiceApi = {
  analyzeSelf: (brandId?: string) => {
    const qs = brandId ? `?brand_id=${brandId}` : "";
    return apiFetch<{ voice_dna: SelfVoiceDNA; message: string }>(
      `/voice/analyze-self${qs}`,
      { method: "POST" }
    );
  },

  getBaseline: (brandId?: string) => {
    const qs = brandId ? `?brand_id=${brandId}` : "";
    return apiFetch<SelfVoiceDNA | null>(`/voice/baseline${qs}`);
  },

  checkDrift: (text: string, brandId?: string) =>
    apiFetch<VoiceDriftResult>("/voice/drift-check", {
      method: "POST",
      body: JSON.stringify({ text, brand_id: brandId }),
    }),
};

File: apps/api/app/schemas/performance.py
"""Pydantic schemas for the Performance Feedback system."""

from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


# ── Create / Update ─────────────────────────────────────────

class ContentPostCreate(BaseModel):
    """Log a published piece of content."""
    title: str = Field(..., min_length=1)
    content_type: str = Field(..., min_length=1)     # youtube_long, linkedin_post, etc.
    platform: str = Field(..., min_length=1)          # youtube, linkedin, instagram, etc.
    hook_used: Optional[str] = None
    hook_type: Optional[str] = None                   # question, bold_claim, story, statistic, contrarian
    topic: Optional[str] = None
    topic_category: Optional[str] = None              # ai_tools, business_strategy, etc.
    content_body: Optional[str] = None
    workflow_id: Optional[str] = None
    collection_id: Optional[str] = None
    published_url: Optional[str] = None
    published_at: Optional[str] = None                # ISO datetime string
    day_of_week: Optional[str] = None
    tags: Optional[List[str]] = None
    brand_id: Optional[str] = None


class ContentPostUpdateMetrics(BaseModel):
    """Update metrics for an existing post (can be partial)."""
    views: Optional[int] = None
    likes: Optional[int] = None
    comments: Optional[int] = None
    shares: Optional[int] = None
    saves: Optional[int] = None
    watch_time_seconds: Optional[int] = None
    click_through_rate: Optional[float] = None
    impressions: Optional[int] = None
    reach: Optional[int] = None
    subscribers_gained: Optional[int] = None


# ── Response Models ─────────────────────────────────────────

class ContentPostSummary(BaseModel):
    """Compact post listing."""
    id: str
    title: str
    content_type: str
    platform: str
    hook_type: Optional[str] = None
    topic: Optional[str] = None
    topic_category: Optional[str] = None
    performance_tier: Optional[str] = None
    engagement_rate: Optional[float] = None
    views: Optional[int] = None
    likes: Optional[int] = None
    comments: Optional[int] = None
    published_at: Optional[str] = None
    created_at: str


class ContentPostDetail(BaseModel):
    """Full post detail including all metrics and AI analysis."""
    id: str
    title: str
    content_type: str
    platform: str
    hook_used: Optional[str] = None
    hook_type: Optional[str] = None
    topic: Optional[str] = None
    topic_category: Optional[str] = None
    content_body: Optional[str] = None
    workflow_id: Optional[str] = None
    collection_id: Optional[str] = None
    published_url: Optional[str] = None
    published_at: Optional[str] = None
    day_of_week: Optional[str] = None
    # Metrics
    views: Optional[int] = None
    likes: Optional[int] = None
    comments: Optional[int] = None
    shares: Optional[int] = None
    saves: Optional[int] = None
    watch_time_seconds: Optional[int] = None
    click_through_rate: Optional[float] = None
    impressions: Optional[int] = None
    reach: Optional[int] = None
    subscribers_gained: Optional[int] = None
    # Calculated
    engagement_rate: Optional[float] = None
    performance_tier: Optional[str] = None
    # AI
    agent_analysis: Dict[str, Any] = {}
    tags: List[str] = []
    metadata: Dict[str, Any] = {}
    created_at: str
    updated_at: str


class PostAnalysisResponse(BaseModel):
    """AI analysis of why a post performed well or poorly."""
    post_id: str
    performance_tier: Optional[str] = None
    analysis: Dict[str, Any] = {}
    message: str


# ── Analytics Models ────────────────────────────────────────

class PlatformBreakdown(BaseModel):
    """Metrics aggregated by platform."""
    platform: str
    post_count: int
    avg_engagement_rate: Optional[float] = None
    avg_views: Optional[float] = None
    top_tier_count: int = 0          # viral + above_average


class TopicBreakdown(BaseModel):
    """Performance by topic category."""
    topic_category: str
    post_count: int
    avg_engagement_rate: Optional[float] = None
    avg_views: Optional[float] = None


class HookBreakdown(BaseModel):
    """Performance by hook type."""
    hook_type: str
    post_count: int
    avg_engagement_rate: Optional[float] = None
    example_hooks: List[str] = []


class PatternDetected(BaseModel):
    """An auto-detected performance pattern."""
    pattern: str               # Human-readable description
    evidence: str              # What data supports this
    confidence: float = 0.5    # 0-1 how confident


class PerformanceAnalytics(BaseModel):
    """Aggregated performance insights."""
    total_posts: int
    platforms: List[PlatformBreakdown] = []
    top_topics: List[TopicBreakdown] = []
    top_hook_types: List[HookBreakdown] = []
    best_day_of_week: Optional[str] = None
    patterns: List[PatternDetected] = []
    top_hooks: List[str] = []          # Actual hook text from best performers
    anti_hooks: List[str] = []         # Hooks that flopped
